### PR TITLE
add mindspore support

### PR DESCRIPTION
**Pull request by lvyufeng**
Add mindspore deep learning framework support.

### CODE REVIEW

**Comment by lvyufeng** [comment]
@arogozhnikov hello sir, I have committed a merge request for adding a new backend named MindSpore, which is a new DL framework. Would please review this pr and give some comments or merge it if no problems?

**Comment by arogozhnikov** [comment]
Hi @lvyufeng 

thanks for a well-organized PR. Github CI does not have cuda, so effectively tests were turned off in your branch.

I've modified it to include tests (but skip einsum) and run on linux. Result is:
- ops are working
- but layers are failing with quite a strange error, 
 
you can see error here in the log: https://github.com/arogozhnikov/einops/runs/7266742212?check_suite_focus=true

here is the branch that includes your code and my change to turn tests on: https://github.com/arogozhnikov/einops/tree/mindspore-show-issue

**Comment by lvyufeng** [comment]
> 

It looks like a bug in CPU platform, I will check it and try to fix it. BTW, I remember the `mindspore-gpu` package i used was a development version, which is a daily compiled version, maybe the bug had beed fixed in MindSpore 1.8(which is released at July 26).

**Comment by lvyufeng** [comment]
it seems that only `Pynative` mode works well, you should add the following code snippet:

```python
from mindspore import context
context.set_context(mode=context.PYNATIVE_MODE)
```

BTW, mindspore will set `Pynative` mode as the default running mode since 1.9.0.

**Comment by lvyufeng** [comment]
the error has been fixed.

**Comment by lvyufeng** [comment]
@arogozhnikov the error has been fixed, would you please review the code and merge it?

File: einops/_backends.py [hunk 1]
```diff
@@ -641,4 +641,55 @@ def layers(self):
...
\ No newline at end of file
+        return self.flow.einsum(pattern, *x)
+
+class MindSporeBackend(AbstractBackend):
+    framework_name = "mindspore"
+
+    def __init__(self):
+        super().__init__()
+        import mindspore as ms
+        ms.set_context(mode=ms.PYNATIVE_MODE)
```
Comment (frelam) [inline einops/_backends.py:652]: Can we use GRAPH_MODE right now ? it seems the very important identity feature in MindSpore.
Thoes who use MindSpore would like to try the jit accelerator.

Comment (lvyufeng) [reply einops/_backends.py:652]: Some pythonic code style or specific grammar is not supported on GRAPH_MODE, like `try...except`, so I have to change the default mode to PYNATIVE. BTW, MindSpore plans to set PYNATIVE mode to support flexible writing, and partial accelerating can use `mindspore.ms_function`.

here is a simple case:
```python
from mindspore import ms_function
from mindspore import nn
class TestNet(nn.Cell):
    def __init__():
        super().__init__():
        self.net = nn.Dense(30, 40)
  
    @ms_function
    def construct(self, x)
        return self.net(x)
```

**Comment by jiangxiluning** [comment]
This feature hasn't been merged yet?

**Comment by arogozhnikov** [comment]
Einops does not accept new backends, those require sufficient time for maintenance.

Lib authors can implement array api:
https://data-apis.org/array-api/latest/API_specification/index.html
and einops.array_api will immediately work for them.

Array api support will be available in einops 0.7

Relevant PR: #261

**Pull request closed without merging**